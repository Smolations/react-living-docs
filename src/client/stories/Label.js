import Story from 'lib/story';
import React from 'react';

import { Card } from './Card';


const globals = {
  React,
  Card,
};

const story = new Story('Label', {
  globals,
});

const _chapterProps = {
  label: 'Totes Labeled',
};

const _chapterCode = `
function LabelInCardHeader(props) {
  return (
    <Card>
      <Card.Head>
        <Label ribbon>{props.label}</Label>
      </Card.Head>

      <Card.Row>
        <p>Sweet header label, eh?</p>
      </Card.Row>
    </Card>
  );
}
`;

story.addChapter({
  id: 'LabelInCardHeader',
  title: 'Label In Card Header',
  props: _chapterProps,
  code: _chapterCode,
});

export default story;
