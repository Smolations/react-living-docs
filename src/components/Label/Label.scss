// unfortunately there is already a .Label class floating around out there so
// we have to differentiate somehow...
.Label2 {
  position: relative;
  display: inline-block;
  margin: $Label-vertical-margin $Label-horizontal-margin;
  padding: $Label-vertical-padding $Label-horizontal-padding;
  background-color: $neutral-tint-3;
  border-radius: 2px;
  color: $neutral-shade-3;
  font-size: 13px;
  font-weight: normal;

  &:first-child {
    margin-left: 0;
  }

  &.Label--blue {
    background-color: $blue;
    color: $white;
  }

  &.Label--green {
    background-color: $green;
    color: $white;
  }

  &.Label--indigo {
    background-color: $indigo;
    color: $white;
  }

  &.Label--orange {
    background-color: $orange;
    color: $white;
  }

  &.Label--red {
    background-color: $red;
    color: $white;
  }

  &.Label--steel {
    background-color: $steel;
    color: $white;
  }

  &.Label--ribbon {
    left: -$Label-horizontal-padding;
    margin-right: 0;

    &.Label--right {
      transform: translateX(-100%);
      left: calc(100% + #{$Label-horizontal-padding});
    }
  }


  /**
   *  integrations
   */
  .CardHead > & {
    &.Label--ribbon {
      left: calc(-#{$Label-horizontal-padding} - #{$CardHead-horizontal-padding});
      margin-right: -$CardHead-horizontal-padding;

      &.Label--right {
        left: calc(100% + #{$Label-horizontal-padding} + #{$CardHead-horizontal-padding});
      }
    }
  }
}
