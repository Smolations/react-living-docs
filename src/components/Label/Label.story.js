import React from 'react';

import { Card } from './Card';


/**
 *  This is a demo of an empty card.
 */
export function LabelInCardHeader(props) {
  return (
    <Card>
      <Card.Head>
        <Label ribbon>{props.label}</Label>
      </Card.Head>

      <Card.Row>
        <Card.Col>
          <p>Sweet header label, eh?</p>
        </Card.Col>
      </Card.Row>
    </Card>
  );
}

LabelInCardHeader.defaultProps = {
  label: 'Totes Labeled',
};
