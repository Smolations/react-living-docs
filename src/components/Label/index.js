/**
 *  `Label`s are a multipurpose component. They can be used as tags, ribbons, and
 *  even form field labels.
 *
 *  @module Label
 */

// eslint-disable-next-line import/prefer-default-export
export { default as Label } from './Label';
