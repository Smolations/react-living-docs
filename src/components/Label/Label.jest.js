import React from 'react';
import assert from 'assert';
import { shallow } from 'enzyme';

import Label from './Label';


describe('Label', () => {
  const labelText = 'labeled';

  it('contains the top-level class name', () => {
    const wrapper = shallow(<Label>{labelText}</Label>);

    assert(wrapper.hasClass('Label2'));
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correct label text', () => {
    const wrapper = shallow(<Label>{labelText}</Label>);
    const renderedLabelText = wrapper.text();

    assert.equal(renderedLabelText, labelText);
    expect(wrapper).toMatchSnapshot();
  });

  it('can be set as a ribbon', () => {
    const wrapper = shallow(<Label ribbon>{labelText}</Label>);

    assert(wrapper.hasClass('Label--ribbon'));
    expect(wrapper).toMatchSnapshot();
  });

  it('can be set as right-positioned', () => {
    const wrapper = shallow(<Label right>{labelText}</Label>);

    assert(wrapper.hasClass('Label--right'));
    expect(wrapper).toMatchSnapshot();
  });

  describe('setting different colors', () => {
    it('can be blue', () => {
      const wrapper = shallow(<Label color="blue">{labelText}</Label>);

      assert(wrapper.hasClass('Label--blue'));
      expect(wrapper).toMatchSnapshot();
    });

    it('can be green', () => {
      const wrapper = shallow(<Label color="green">{labelText}</Label>);

      assert(wrapper.hasClass('Label--green'));
      expect(wrapper).toMatchSnapshot();
    });

    it('can be indigo', () => {
      const wrapper = shallow(<Label color="indigo">{labelText}</Label>);

      assert(wrapper.hasClass('Label--indigo'));
      expect(wrapper).toMatchSnapshot();
    });

    it('can be orange', () => {
      const wrapper = shallow(<Label color="orange">{labelText}</Label>);

      assert(wrapper.hasClass('Label--orange'));
      expect(wrapper).toMatchSnapshot();
    });

    it('can be red', () => {
      const wrapper = shallow(<Label color="red">{labelText}</Label>);

      assert(wrapper.hasClass('Label--red'));
      expect(wrapper).toMatchSnapshot();
    });

    it('can be steel', () => {
      const wrapper = shallow(<Label color="steel">{labelText}</Label>);

      assert(wrapper.hasClass('Label--steel'));
      expect(wrapper).toMatchSnapshot();
    });
  });
});
