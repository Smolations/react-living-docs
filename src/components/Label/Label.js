import PropTypes from 'prop-types';
import React from 'react';

import './Label.scss';


const colorClassMap = {
  blue: 'Label--blue',
  green: 'Label--green',
  indigo: 'Label--indigo',
  orange: 'Label--orange',
  red: 'Label--red',
  steel: 'Label--steel',
};


/**
 *  The Label component.
 *
 *  @param {string}  color  Change the color scheme of the label. Supported colors are:
 *                          blue, green, indigo, orange, red, steel.
 *  @param {boolean} ribbon Make the label display as a ribbon, which overlaps the edge
 *                          of it's parent container. The parent container should be
 *                          styled as `position: relative`.
 *  @param {boolean} right  If the `Label` is designated as a `ribbon`, make it hang off
 *                          the right side of the parent container instead of the left.
 *                          NOTE: Be sure the `Label` element is the last child of the
 *                          container in order to render properly.
 */
export default function Label(props) {
  const {
    children,
    color,
    ribbon,
    right,
  } = props;

  // temporary name since .Label is taken
  const classes = ['Label2'];

  if (color && colorClassMap[color]) {
    classes.push(colorClassMap[color]);
  }

  ribbon && classes.push('Label--ribbon');
  right && classes.push('Label--right');

  return (
    <div className={classes.join(' ')}>
      {children}
    </div>
  );
}


Label.propTypes = {
  children: PropTypes.node.isRequired,
  color: PropTypes.string,
  ribbon: PropTypes.bool,
  right: PropTypes.bool,
};
