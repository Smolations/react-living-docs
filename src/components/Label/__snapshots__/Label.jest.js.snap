// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label can be set as a ribbon 1`] = `
<div
  className="Label2 Label--ribbon"
>
  labeled
</div>
`;

exports[`Label can be set as right-positioned 1`] = `
<div
  className="Label2 Label--right"
>
  labeled
</div>
`;

exports[`Label contains the top-level class name 1`] = `
<div
  className="Label2"
>
  labeled
</div>
`;

exports[`Label renders correct label text 1`] = `
<div
  className="Label2"
>
  labeled
</div>
`;

exports[`Label setting different colors can be blue 1`] = `
<div
  className="Label2 Label--blue"
>
  labeled
</div>
`;

exports[`Label setting different colors can be green 1`] = `
<div
  className="Label2 Label--green"
>
  labeled
</div>
`;

exports[`Label setting different colors can be indigo 1`] = `
<div
  className="Label2 Label--indigo"
>
  labeled
</div>
`;

exports[`Label setting different colors can be orange 1`] = `
<div
  className="Label2 Label--orange"
>
  labeled
</div>
`;

exports[`Label setting different colors can be red 1`] = `
<div
  className="Label2 Label--red"
>
  labeled
</div>
`;

exports[`Label setting different colors can be steel 1`] = `
<div
  className="Label2 Label--steel"
>
  labeled
</div>
`;
